### ISSUE TITLE
Docker Daemon crashes after starting lots of containers

### ISSUE THREAD

**Issue by jpbochi** [docker/for-mac#826]
### Expected behavior

Docker should continue running.
### Actual behavior

A couple of minutes after starting some 15 containers using docker-compose, docker daemon crashes.
### Information

Output from Docker for Mac Diagnose & Feeback tool:

```
Docker for Mac: version: 1.12.1 (2d5b4d9)
OS X: version 10.11.6 (build: 15G1004)
logs: /tmp/AE12FEE1-78CE-496A-A738-6F2C6944CFB5/20161019-170956.tar.gz

[OK]     docker-cli
[OK]     virtualization kern.hv_support
[OK]     menubar
[OK]     moby-syslog
[OK]     dns
[OK]     disk
[OK]     system
[OK]     app
[OK]     osxfs
[OK]     virtualization VT-X
[OK]     db
[OK]     slirp
[OK]     logs
[OK]     env
[OK]     vmnetd
[OK]     moby-console
[OK]     moby
[OK]     driver.amd64-linux
```

[20161019-170956.tar.gz](https://github.com/docker/for-mac/files/539412/20161019-170956.tar.gz)

```
$ docker-compose version
docker-compose version 1.8.0, build f3628c7
docker-py version: 1.9.0
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.2h  3 May 2016
```

Output of `docker-compose up` when the crash happens:

```
Exception in thread Thread-17:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 502 Server Error: Bad Gateway ("Bad response from Docker engine")

Exception in thread Thread-29:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-16:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-20:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-22:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-26:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-24:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-30:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-23:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-27:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-28:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-19:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-18:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-21:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")

Exception in thread Thread-25:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 500 Server Error: Internal Server Error ("dial unix /Users/joao0191/Library/Containers/com.docker.docker/Data/*00000003.00000948: connect: connection refused")
```
### Steps to reproduce the behavior
1. `docker-compose up` starting lots of containers, with mounts to local FS, links between them, and some ports being exposed. Unfortunately, I can't share the `docker-compose.yml` files I've been using.
2. After running an script that connects to ports being exposed by some of the created containers, it takes a few seconds for my docker daemon to crash.

**Comment by rogaha** [comment]
Detected symptom of problem 'aufs-corruption' in AE12FEE1-78CE-496A-A738-6F2C6944CFB5/20161019-170956.

The daemon cannot start because some files in /var/lib/docker are corrupt. This can happen if the filesystem is unmounted uncleanly. The workaround is to reset to factory defaults which will unfortunately delete /var/lib/docker: all containers and images will need to be rebuilt.

May be related to docker/for-mac#20

**Comment by rogaha** [comment]
@jpbochi can you please try that approach and confirm that fixes the issue?

**Comment by jpbochi** [comment]
@rogaha I saw this problem before and did a factory reset already. It came back soon afterwards.

Maybe the root cause is some other thing I'm doing that is causing the FS corruption. I'll do another factory reset and keep an eye out for that. Thanks.

**Comment by ijc25** [comment]
This issue has been inactive for more than 14 days while marked as `status/0-more-info-needed`. It is being closed due to abandonment. Please feel free to re-open with more information about the problem. 

MORE_INFO_EXPIRY_TIMEOUT

**Comment by ijc25** [comment]
(apologies for the previous comment, I pressed the wrong button in our triage tool).

@jpbochi prior to the aufs corruption your original report your logs have a bunch of
```
<Error>: WATCH SUBSCRIBER REMOVAL INVARIANT VIOLATED
```
messages. This was a known issue which was fixed in stable with the 1.12.3 release. I believe it can also cause crashes which would explain your subsequent aufs issue. Upgrading to 1.12.3 should therefore fix both of these for you. You may need to do one final reset to factory defaults in order to flush any residual issues from before the upgrade.

I'll close this now, please reopen if you continue to see issues.

**Comment by jpbochi** [comment]
@ijc25 Thanks for the update. I did realise at some point that the containers were not really crashing, and it was docker-compose that was failing to follow their logs. I did update to 1.12.3 a few days ago (I think) and I don't see the problem anymore. Thanks for the fix.
